### PR TITLE
buildDunePackage: use dune_2 by default

### PR DESCRIPTION
**Pull request by vbgl**
###### Description of changes

Borrows some changes from #162745 and the purpose of #163766.

cc @ulrikstrid @sternenseemann 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by ulrikstrid — APPROVED** [review]
Looks good to me, we still want the non-dune2 PR but I can fix it up after this lands

File: pkgs/development/ocaml-modules/spacetime_lib/default.nix [hunk 1]
```diff
@@ -1,4 +1,7 @@
-{ lib, fetchFromGitHub, buildDunePackage, owee }:
+{ lib, fetchFromGitHub, buildDunePackage, ocaml, owee }:
+
+lib.throwIfNot (lib.versionAtLeast "4.12" ocaml.version)
```
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/spacetime_lib/default.nix:3]: Wouldn't this be possible using `minimalOCamlVersion`?

**Review by sternenseemann — COMMENTED** [review]
According to a quick grep `pkgs/development/ocaml-modules/reason-native/console.nix` also relies on the default value, did you check whether it still works with dune 2?

Other than that, the change seems to be in order, although I did not do any builds to check for regressions.

File: pkgs/development/ocaml-modules/spacetime_lib/default.nix (hunk 1 above)
Comment (vbgl) [reply pkgs/development/ocaml-modules/spacetime_lib/default.nix:3]: No. Minimal sets a lower bound. Here it’s a upper bound.

**Comment by vbgl** [comment]
> According to a quick grep `pkgs/development/ocaml-modules/reason-native/console.nix` also relies on the default value, did you check whether it still works with dune 2?

I did rebuild all changed path. This includes `reason-native` for OCaml versions from 4.04 to 4.07 (included).

File: pkgs/development/ocaml-modules/spacetime_lib/default.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/ocaml-modules/spacetime_lib/default.nix:3]: Right, I feel like `lib.versionOlder ocaml.version "4.13"` would be a bit more readable, but not too important.

**Review by sternenseemann — APPROVED** [review]

**Comment by sternenseemann** [comment]
Hm, a small note in the 22.05 changelog would be good as well, I suppose?

**Pull request merged**